lib/phoenix_live_view/helpers.ex

```diff
@@ -863,27 +902,70 @@ defmodule Phoenix.LiveView.Helpers do
 
   ## Examples
 
-      <.form_for let={f} data={@changeset}>
+      <.form let={f} for={@changeset}>
         <%= text_input f, :name %>
-      </.form_for>
+      </.form>
 
-      <.form_for let={user_form} data={@changeset} as="user">
+      <.form let={user_form} for={@changeset} as="user" {@extra}>
         <%= text_input user_form, :name %>
-      </.form_for>
+      </.form>
   """
-  def form_for(assigns) do
-    opts =
-      assigns |> Map.take([:as, :method, :multipart, :csrf_token, :errors, :id]) |> Enum.into([])
+  @form_opts [:as, :method, :multipart, :csrf_token, :errors, :id]
+  def form(assigns) do
+    action = assigns[:action] || "#"
+
+    {form_for, assigns} =
+      Map.pop_lazy(assigns, :for, fn -> raise ArgumentError, "missing :for assign to form" end)
+
+    {csrf_token, assigns} =
+      Map.pop_lazy(assigns, :csrf_token, fn -> Phoenix.HTML.Tag.csrf_token_value(action) end)
+
+    form_opts = assigns |> Map.take(@form_opts) |> Enum.into([])
+
+    form = %Phoenix.HTML.Form{Phoenix.HTML.FormData.to_form(form_for, form_opts) | action: action}
+
+    {method, hidden_method} = form_method(assigns[:method])
 
     assigns =
       assigns
-      |> Map.put(:opts, opts)
-      |> Map.put(:url, assigns[:url] || "#")
+      |> LiveView.assign(form: form, id: form.id, action: action)
+      |> LiveView.assign(enctype: enctype(assigns))
+      |> LiveView.assign(csrf_token: csrf_token)
+      |> LiveView.assign(method: method, hidden_method: hidden_method)
+      |> LiveView.assign(attrs: assigns_to_attrs(assigns, [:for, :action | @form_opts]))
```

Comment:
Convert this into a single `assign/2` call and it should be quite more efficient. :)

```suggestion
      |> LiveView.assign(form: form, id: form.id, action: action)
      |> LiveView.assign(enctype: enctype(assigns))
      |> LiveView.assign(csrf_token: csrf_token)
      |> LiveView.assign(method: method, hidden_method: hidden_method)
      |> LiveView.assign(attrs: assigns_to_attrs(assigns, [:for, :action | @form_opts]))
```